zone/bonuses.cpp

```diff
@@ -128,54 +130,87 @@ void Client::CalcBonuses()
 		consume_food_timer.SetTimer(timer);
 }
 
-int Mob::CalcRecommendedLevelBonus(uint8 level, uint8 reclevel, int basestat)
+int Mob::CalcRecommendedLevelBonus(uint8 current_level, uint8 recommended_level, int base_stat)
 {
-	if( (reclevel > 0) && (level < reclevel) )
-	{
-		int32 statmod = (level * 10000 / reclevel) * basestat;
+	if (recommended_level && current_level < recommended_level) {
+		int32 stat_modifier = (current_level * 10000 / recommended_level) * base_stat;
 
-		if( statmod < 0 )
-		{
-			statmod -= 5000;
-			return (statmod/10000);
-		}
-		else
-		{
-			statmod += 5000;
-			return (statmod/10000);
-		}
+		stat_modifier += stat_modifier < 0 ? -5000 : 5000;
+
+		return (stat_modifier / 10000);
 	}
 
 	return 0;
 }
 
-void Client::CalcItemBonuses(StatBonuses* newbon) {
-	//memset assumed to be done by caller.
 
-	// Clear item faction mods
+
+// These item stat caps depend on spells/AAs so we process them after those are processed
+void Client::ProcessItemCaps()
+{
+	itembonuses.HPRegen = std::min(itembonuses.HPRegen, CalcHPRegenCap());
+	itembonuses.ManaRegen = std::min(itembonuses.ManaRegen, CalcManaRegenCap());
+	itembonuses.EnduranceRegen = std::min(itembonuses.EnduranceRegen, CalcEnduranceRegenCap());
+
+	// The Sleeper Tomb Avatar proc counts towards item ATK
+	// The client uses a 100 here, so using a 100 here the client and server will agree
+	// For example, if you set the effect to be 200 it will get 100 item ATK and 100 spell ATK
+	if (IsValidSpell(SPELL_AVATAR_ST_PROC) && FindBuff(SPELL_AVATAR_ST_PROC)) {
+		itembonuses.ATK += 100;
+		spellbonuses.ATK -= 100;
+	}
+
+	itembonuses.ATK = std::min(itembonuses.ATK, CalcItemATKCap());
+}
+
+void Mob::CalcItemBonuses(StatBonuses* b) {
 	ClearItemFactionBonuses();
 	SetShieldEquiped(false);
 	SetTwoHandBluntEquiped(false);
 	SetTwoHanderEquipped(false);
 	SetDuelWeaponsEquiped(false);
 
-	unsigned int i;
-	// Update: MainAmmo should only calc skill mods (TODO: Check for other cases)
+	int16 i;
+
 	for (i = EQ::invslot::BONUS_BEGIN; i <= EQ::invslot::BONUS_SKILL_END; i++) {
-		const EQ::ItemInstance* inst = m_inv[i];
-		if(inst == 0)
+		const EQ::ItemInstance* inst = nullptr;
+		if (IsOfClientBotMerc()) {
+			inst = GetInv().GetItem(i);
+		} else {
```

Comment:
One could assume the only thing that falls under this else is NPC but we should be explicit in an else if IsNPC()